### PR TITLE
Decode File URI for local iOS files

### PR DESCRIPTION
**Pull request by barun1997**
### Fixed Issues

$ https://github.com/Expensify/App/issues/7962

### Tests

- [ ] Upload a PDF file with a space in naming (such as `Sample File.pdf`)
- [ ] Check if you can see a preview for the PDF

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. “toggleReport” and not “onIconClick”)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct english, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct english and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY
- [x] I verified any variables that can be defined as constants (ie. in CONST.js) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] Any internal methods are bound to “this” properly so there are no scoping issues
    - [ ] Any internal methods bound to “this” are necessary to be bound
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function
(i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)

#### PR Reviewer Checklist
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. “toggleReport” and not “onIconClick”).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct english, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct english and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components are not impacted by changes in this PR (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] Any internal methods are bound to “this” properly so there are no scoping issues
    - [ ] Any internal methods bound to “this” are necessary to be bound
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function
(i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)

### QA Steps

- [ ] Upload a PDF file with a space in naming (such as `Sample File.pdf`)
- [ ] Check if you can see a preview for the PDF

#### Web

Screen Shot 2022-03-20 at 7 19 18 AM

#### Mobile Web

Screen Shot 2022-03-20 at 7 20 02 AM

#### Desktop

Screen Shot 2022-03-20 at 7 19 42 AM

#### iOS

https://user-images.githubusercontent.com/19339044/159144226-5790a66d-bc7e-4282-82d6-38a59ab7d5ba.mp4

#### Android

Screen Shot 2022-03-20 at 7 20 36 AM

### CODE REVIEW

File: src/libs/getLocalFileUri/index.js
Comment (parasharrajat) [inline src/libs/getLocalFileUri/index.js:7]: Why can't we just use `decodeURIComponent` everywhere?

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/libs/getLocalFileUri/index.js
Comment (barun1997) [reply src/libs/getLocalFileUri/index.js]: @parasharrajat I believe I changed it because I was decoding the URL in another location. You are correct though, we can directly `decodeURIComponent` here. I made the change accordingly!

**Comment by parasharrajat** [comment]
Testing

**Review by parasharrajat — CHANGES_REQUESTED** [review]
There is an issue when we file name has special chars e.g `dummy%201.pdf`. Create a file with this name and try to add it to chat.

**Comment by barun1997** [comment]
> There is an issue when we file name has special chars e.g `dummy%201.pdf`. Create a file with this name and try to add it to chat.

Looking into this!

**Comment by barun1997** [comment]
> There is an issue when we file name has special chars e.g `dummy%201.pdf`. Create a file with this name and try to add it to chat.

Hi @parasharrajat, could you tell me which environment this is in? Seems to work in iOS for me. 

https://user-images.githubusercontent.com/19339044/159414306-dc921c99-f6ca-4106-b34c-9cd72448d943.mp4

**Comment by parasharrajat** [comment]
Android

**Comment by barun1997** [comment]
> Android

Ah okay! Got it. Thanks for letting me know!

**Comment by barun1997** [comment]
hi @parasharrajat, please feel free to review it again.

**Comment by barun1997** [comment]
@parasharrajat, Not sure I understand. Could you please elaborate?

**Comment by parasharrajat** [comment]
I meant to say that It would be a great help to the RN community if we fix this issue upstream on the [react-native-document-picker](https://www.npmjs.com/package/react-native-document-picker).

**Comment by barun1997** [comment]
Oh yeah. @parasharrajat. Looks like the contributor is aware of the problem. https://github.com/rnmods/react-native-document-picker/issues/199#issuecomment-932807046

**Comment by parasharrajat** [comment]
It seems the maintainer is not convinced to fix it. So someone has to drop a PR to fix it. @barun1997 Would you be able to send a PR upstream?

**Comment by barun1997** [comment]
@parasharrajat I sent a PR to fix it. Waiting for the maintainer to take a look.

**Comment by parasharrajat** [comment]
Thanks, could you please share the link?

**Comment by roryabraham** [comment]
https://github.com/rnmods/react-native-document-picker/pull/544

**Comment by parasharrajat** [comment]
Let's wait for a few days.

**Comment by roryabraham** [comment]
@barun1997 I commented on the PR with a minor improvement.

**Comment by roryabraham** [comment]
@barun1997 I also suggest that you run `npm run test` and `npm run lint` in that repo so that we can be more confident tests will pass if/when the contributor runs PR checks.

**Comment by barun1997** [comment]
@roryabraham Thanks for the suggestion. 

I did! There were no tests and no problems with lint. Fingers crossed🤞 now 😅

File: src/libs/getLocalFileUri/index.ios.js
Comment (parasharrajat) [inline src/libs/getLocalFileUri/index.ios.js:2]:
```suggestion
 * Returns the decoded URI String. React-native-document-picker returns encoded file path on iOS.
```

File: src/libs/getLocalFileUri/index.ios.js
Comment (parasharrajat) [inline src/libs/getLocalFileUri/index.ios.js:3]:
```suggestion
 * @param {String} localFileUri
```

File: src/components/AttachmentModal.js
Comment (parasharrajat) [inline src/components/AttachmentModal.js:21]:
```suggestion
import getLocalFileUri from '../libs/getLocalFileUri';
```

Rename this to getDecodedURIPath and the rename the file to same

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by parasharrajat** [comment]
There is no update on the PR upstream. @barun1997 Did you follow up on the PR?

**Comment by barun1997** [comment]
@parasharrajat  made the change!

**Comment by barun1997** [comment]
The maintainer has mentioned he is not in favor of changing the behavior in our proposed way, so we might have to do with either) 1 - forking the PR as @roryabraham mentioned or 2 - mark the implementation in my PR as permanent solution

**Comment by parasharrajat** [comment]
Ok. I also feel that those changes should be done on native code.

**Comment by parasharrajat** [comment]
Well, I was debugging this issue in detail. And it seems we were looking at the wrong place to fix the bug. The issue is with react-native-pdf lib instead of react-native-document-picker. 

During debugging I found that on both Android and iOS file path received after uploading the file is encoded so our fix is not correct logically.  

Looking at the code, it seems that on [Android](https://github.com/wonday/react-native-pdf/blob/073161045b9d6080714e685a685b86e1f5963fcf/android/src/main/java/org/wonday/pdf/PdfView.java#L242)  it parses the file path but not on [ios](https://github.com/wonday/react-native-pdf/blob/073161045b9d6080714e685a685b86e1f5963fcf/ios/RCTPdf/RCTPdfView.m#L135) which is the whole issue. 

So we should fix this on react-native-pdf instead.

**Comment by roryabraham** [comment]
> The maintainer has mentioned he is not in favor of changing the behavior in our proposed way, so we might have to do with either) 1 - forking the PR as @roryabraham mentioned
I think we can close the PR you created in react-native-document-picker. In general we want to minimize our use of forks as much as possible. When we _do_ use them, we want it to be as temporary as possible, while we're waiting for a PR with our fixes to be reviewed and merged in the upstream. In this case, the maintainer has provided a detailed explanation for why they don't want to merge the PR, so we know they won't be merging the code into the upstream. Therefore, we _do not_ want to work with a fork.

> or mark the implementation in my PR as permanent solution
I haven't worked too closely on this issue, so I'll defer to @parasharrajat's judgement on this in the meantime. Let me know if you reach an impasse or want my perspective again.

**Comment by barun1997** [comment]
> Looking at the code, it seems that on [Android](https://github.com/wonday/react-native-pdf/blob/073161045b9d6080714e685a685b86e1f5963fcf/android/src/main/java/org/wonday/pdf/PdfView.java#L242) it parses the file path but not on [ios](https://github.com/wonday/react-native-pdf/blob/073161045b9d6080714e685a685b86e1f5963fcf/ios/RCTPdf/RCTPdfView.m#L135) which is the whole issue. I may be wrong about it(due to native code). But I think this is the case.
> So we should fix this on react-native-pdf instead.

I will try drilling it too, but I am not familiar with native code as well!

**Comment by parasharrajat** [comment]
Steps would be to create a reproduction using the react-native-document-picker and react-native-pdf and see if the URI value is encoded on both IOS and Android. Then Whether RN-pdf renders the pdf on both platforms or not.  

If we see that RN-PDF has the mentioned issue and then open a new issue on their repo and wait for response.

**Comment by parasharrajat** [comment]
@barun1997 Let us know if you are facing any challanges.

**Comment by barun1997** [comment]
@parasharrajat I sent an issue in react-pdf.

File: src/libs/getDecodedURIFromDocumentPicker/index.ios.js
Comment (parasharrajat) [inline src/libs/getDecodedURIFromDocumentPicker/index.ios.js:2]: Let's mention in this comment that this issue exists with react-native-pdf
```suggestion
 * Returns the decoded URI String. react-native-pdf fails to parse the files with spaces on iOS.
```
```

```

File: src/libs/getDecodedURIFromDocumentPicker/index.js
Comment (parasharrajat) [inline src/libs/getDecodedURIFromDocumentPicker/index.js:1]: Let's call this lib `normalizeFileURI`

**Review by parasharrajat — CHANGES_REQUESTED** [review]
It is been many days without any update on the upstream repo `react-native-pdf`. So I think it is time to move along with our fix and resolve the issue on the App. 

Please merge main

**Comment by roryabraham** [comment]
> It is been many days without any update on the upstream repo react-native-pdf. So I think it is time to move along with our fix and resolve the issue on the App.

Not sure I agree with this – if we've identified that the bug exists in react-native-pdf, that's where we should solve it, especially since the bug isn't very urgent.

**Comment by roryabraham** [comment]
Commented about this in the linked issue https://github.com/Expensify/App/issues/7962#issuecomment-1122846471

**Comment by roryabraham** [comment]
Going to close this out for now

**Pull request closed without merging**
